src/plugins/intel_cpu/src/graph.cpp

```diff
@@ -804,9 +1008,15 @@ void Graph::AllocateWithReuse(const std::vector<size_t>& syncNodesInds) {
 
         int64_t boxSize = 0;
         bool isConst = false, isOutput = false, isInput = false;
-        for (auto &edge : edge_clusters[i]) {
-            int e_start = edge->getParent()->getExecIndex();
-            int e_finish = edge->getChild()->getExecIndex();
+
+        for (auto &edge : clusters[i]) {
+            const auto& parent = edge->getParent();
+            const auto& child = edge->getChild();
+
+            // If node uses its input / output memory multiple times in scope of a single execution (i.e TensorIterator)
+            // prolong the lifetime of a memory region till execution is finished
+            int e_start = parent->usesInOutMemoryMultipleTimes() ?  globalExecIndex.at(parent).first : globalExecIndex.at(parent).second;
+            int e_finish = child->usesInOutMemoryMultipleTimes() ?  globalExecIndex.at(child).second : globalExecIndex.at(child).first;
```

Comment:
The type check is not enough.
In proper implementation we will need to know a memory from which port is actually being used multiple times (see the comment in `usesInOutMemoryMultipleTimes`)
Edge modifier seems even more excessive, I mean to add a boolean flag to data structure just for one a single case.

Reply:
It's unlikely that we would have loop like nodes other than `TensorIterator`, so instead of modifying the `Node` interface, it just enough to downcast and process exactly tensor iterator changing the interface of only that class.

Reply:
Sure. Let's to it this way